### ISSUE TITLE
Error in training

### ISSUE THREAD

**Issue by TNA8** [AhmetSencan/MaskSplit-Self-supervised-Meta-learning-for-Few-shot-Semantic-Segmentation#2]
I downloaded the pretrained backbones for resnet101 and used `pascal/split0/best.pth` in the zip.
When I try to train the model on pascal dataset, I faced this error.
![screenshot_92](https://user-images.githubusercontent.com/58914448/145524405-44870253-197e-4829-a37d-810a27c60f7c.png)

**Comment by AhmetSencan** [comment]
Hi @TNA8,
I checked the link I provided for the pre-trained backbones. Apparently it was a wrong link. Now I updated the README.md and you can download resnet101_v2.pth from the correct link. 

Sorry for any inconvenience.

**Comment by AhmetSencan** [comment]
I was able to recreate the issue. I think it is caused by a recent update to pythorch lightning, which our code is apparently not compatible with. Please use the version given in requirements. We will look into the changes.

**Comment by TNA8** [comment]
* I installed correct version of dependencies in README.
* Downloaded the correct pretrained resnet101 model.

When I try to run the training, I faced another error.
![screenshot_98](https://user-images.githubusercontent.com/58914448/145797947-e8d6f292-2385-4413-9699-cc776c6e8f26.png)

**Comment by AhmetSencan** [comment]
I was able to replicate this error by giving a different data path in the config file. In config file `data_root:` should be assigned to the directory that contains JPEGImages, SegmentationClassAug and saliency_unsupervised_model. For example in my case that line is as follows: `data_root: /home/ahmet/data/pascal/`. I checked the code and I can see that it is not clearly stated. I will also update the files.

**Comment by TNA8** [comment]
I correctly assigned the `data_root`. I will be waiting your update. Thanks.

**Comment by TNA8** [comment]
I successfully configured for one class dataset and started training.
Could you explain how to analyze the training log?
* Why do we have several `val_miou`s?
* Why are some `val_miou`s `nan`?
![screenshot_104](https://user-images.githubusercontent.com/58914448/146341215-02675475-5663-45ce-a59c-d2194d3faaba.png)

**Comment by sercanamac** [comment]
Hello @TNA8 there are several mious because you are training with "unsupervised_all" mode. Since we assume that we do not have any labels, it means that we can evaluate our model on all folds at once. So val_miou_0 means validation result for fold1, and val_miou_1 means that validation result for fold2 and so on... I don't know why you are getting NaNs though, I think we could train our models without any problem.

**Comment by AhmetSencan** [comment]
For the first question: 
I think in your configuration file, the `use_all_classes:` is `True`. That is why in the simpleNetwork.py, the code
```
        if not self.args.use_all_classes:
            self.val_class_IoU = [ClassIoUNew(self.args.num_classes_val)]
        else:
            self.val_class_IoU = [ClassIoUNew(self.args.num_classes_val),
                                  ClassIoUNew(self.args.num_classes_val),
                                  ClassIoUNew(self.args.num_classes_val),
                                  ClassIoUNew(self.args.num_classes_val)]
```
creates 4 `ClassIoUNew` objects. This is used when we are training PASCAL, where training procedure is the same, but for 4 different folds, we do 4 separate validations. In your case, there is only one fold with one class. Hence you can either set `use_all_classes:` to `False` or you can safely ignore `val_miou_1`, `val_miou_2`, `val_miou_3`. 

`val_miou_old` is a old version of miou calculation. To follow the literature on the topic, we implemented `ClassIoUNew`. But `val_miou_old` remains for debugging and sanity check purposes. 

The one that should be followed is `val_miou_0` in your case.

For the second question: 
When we are computing the mIoU, we use the compute method in the `ClassIoUNew` class. If you check the code, you see that in the first line of this, there is the code snippet `print(self.cls_iou, self.cls_counts)`. 
In the picture, these are shown as 
```
tensor([1.0964]) tensor([64.])
tensor([0.]) tensor([0.])
tensor([0.]) tensor([0.])
tensor([0.]) tensor([0.])
```
The first line is for `val_miou_0`, second is for `val_miou_1` and so on. In the second line of `compute` method we divide `self.cls_iou` to `self.cls_counts`. Since you have only one class, which is in the first fold, you are never updating the remaining  `ClassIoUNew` objects and self.cls_counts always remains zero and this is the reason for `nan` values. Since you only have to keep track of `val_miou_0`, these `nan` values can safely be ignored

**Comment by TNA8** [comment]
In my case, `val_miou_0` is always zero. Do you have any idea about it?

**Comment by sercanamac** [comment]
Closed as we don't know about OP's custom data. In principle, you don't really need to train another model for your own class. Our model is supposed to be class agnostic and you can just run the inference script with a support and query image of your class. If you want to train your own model we suggest collecting a big dataset and do any modifications you may need.
